src/rgw/rgw_op.h

```diff
@@ -120,6 +120,37 @@ class RGWOp {
   int do_aws4_auth_completion();
 
   virtual int init_quota();
+
+  // So! Now and then when we try to update bucket information, the
+  // bucket has changed during the course of the operation. (Or we
+  // have a cache consistency problem that Watch/Notify isn't ruling
+  // out completely.)
+  //
+  // When this happens, we need to update the bucket info and try
+  // again. It's important to know what *part* to try again,
+  // however. We can't simply re-send, since that will obliterate the
+  // update.
+  //
+  // Thus, callers of this function should include everything that
+  // merges information to be changed into the bucket information as
+  // well as the set.
+  //
+  // The called function should set op_ret, similar to code in execute()
+  template<typename F>
+  void retry_raced_bucket_write(F&& f) {
```

Comment:
@adamemerson RGWOp is not the place for a bucket specific operation. I suggest that you move it to a function in rgw_op.cc, make it return the ret value, and have f() return op_ret.